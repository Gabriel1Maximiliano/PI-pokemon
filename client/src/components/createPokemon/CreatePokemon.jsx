import React from 'react'
import { useFormulario } from '../../hooks/useFormulario';
import { BarraDeNavegacion } from '../navegacion/BarraDeNavegacion';
import './createPokemon.css'

export const CreatePokemon = () => {

  const initialForm = {
    name: "",
    height: "",
    weight: "",
    hp: "",
    strength: "",
    defense: "",
    speed: "",
    type: [],
    img: ""
  };
  const validationForm = (form) => {
    let errors = {};
    let regexURL = /(^(https:\/\/).*\.(png|jpg))/
    let regexName = /^[A-Za-z0-9ÑñÁáÉéÍíÓóÚúÜü\s]+$/;
    if (!form.name.trim()) {
      errors.name = 'El campo nombre es requerido'
    }
    else if (!regexName.test(form.name.trim())) {
      errors.name = 'Caracter inválido solo letras y números'
    }

    if (!form.height) {
      errors.height = 'Agregue altura a su Pokemon'
    } else if (form.height < 1) {
      errors.height = 'Altura de al menos 1'
    } else if (form.height > 100) { //1-100
      errors.height = 'La altura debe ser menor o igual a 100'
    }
    if (!form.weight) {
      errors.weight = 'Agregue peso a su Pokemon'
    }
    else if (form.weight < 1) {
      errors.weight = 'Peso de al menos 1 Kg'
    } else if (form.weight > 100) { //1-100
      errors.weight = 'La vida debe ser menor o igual a 100'
    }
    if (!form.hp) {
      errors.hp = 'Agregue puntos de vida a su Pokemon'
    } else if (form.hp < 1) {
      errors.hp = 'Vida de al menos 1'
    } else if (form.hp > 100) { //1-100
      errors.hp = 'La vida debe ser menor o igual a 100'
    }
    if (!form.strength) {
      errors.strength = 'Agregue fuerza de ataque a tu Pokemon'
    } else if (form.strength < 1) {
      errors.strength = 'Ataque de al menos 1'
    }
    if (!form.defense) {
      errors.defense = 'Agregue defensa a tu Pokeon'
    } else if (form.defense < 1) {
  
      errors.defense = 'Defensa de al menos 1'
    } else if (form.defense > 100) { //1-100
      errors.defense = 'La defensa debe ser menor o igual a 100'
    }
    if (!form.speed) {
      errors.speed = 'Agregue velocidad a tu Pokemon'
    } else if (form.speed < 1) {
   
      errors.speed = 'Velocidad de al menos 1'
    } else if (form.speed > 100) { //1-100
      errors.speed = 'La velocidad debe ser menor o igual a 100'
    }
    if (form.type.lenght > 0) {
      errors.type = 'Selecione un tipo de Pokemon'
     
    }
    
    console.log(form.img)
    if(!regexURL.test(form.img)){

      errors.img = 'Formato de imagen no válido'
    }
    
    
    return errors
  }
  const { form, errors, handleBlur, handleChange, handleSubmit, handleEliminate } = useFormulario(initialForm, validationForm);



  return (
    <div className='contenedor-formulario' >
      <br />
      <BarraDeNavegacion />
     
      <br></br>
      <div className='formulario' >
        <form onSubmit={handleSubmit} >
          <label htmlFor='name' >Nombre del Pokemon</label>
          <br></br>
          <br></br>
          <div className='decoration-input' >
            <input
              className='input'
              type='text'
              name='name'
              id='name'
              value={form.name}
              placeholder='Ingrese nombre del Pokemon'
              onChange={handleChange}
              onBlur={handleBlur}
            />

            {
              errors.name && <p style={{ color: 'green' }} >{errors.name}</p>
            }
            <br></br>
            <br></br>
            <label htmlFor='height' >Altura entre {`(1-30)`} </label>
            <input
              type='number'
              name='height'
              min="1"
              max="30"
              id='height'
              onChange={handleChange}
              value={form.height}
              onBlur={handleBlur}
            />
             {
              errors.height && <p style={{ color: 'green' }} >{errors.height}</p>
            }
            <br></br>
            <br></br>
            <label htmlFor='weight' >Peso entre {`(1-300)`} </label>
            <input
              type='number'
              name='weight'
              min="1"
              max="300"
              id='weight'
              onChange={handleChange}
              value={form.weight}
              onBlur={handleBlur}
            />
            {
              errors.weight && <p style={{ color: 'green' }} >{errors.weight}</p>
            }
            <br></br>
            <br></br>
            <label htmlFor='hp' >Vida entre {`(1-100)`} </label>
            <input
              type='number'
              name='hp'
              min="1"
              max="100"
              id='hp'
              onChange={handleChange}
              value={form.hp}
              onBlur={handleBlur}
            />
            {
              errors.hp && <p style={{ color: 'green' }} >{errors.hp}</p>
            }
            <br></br>
            <br></br>
            <label htmlFor='strength' >Ataque entre {`(1-300)`} </label>
            <input
              type='number'
              name='strength'
              min="1"
              max="300"
              id='strength'
              onChange={handleChange}
              value={form.strength}
              onBlur={handleBlur}
            />
            {
              errors.strength && <p style={{ color: 'green' }} >{errors.strength}</p>
            }
            <br></br>
            <br></br>
            <label htmlFor='defense'>Defensa entre {`(1-100)`} </label>
            <input
              type='number'
              name='defense'

              min="1"
              max="100"
              id='defense'
              onChange={handleChange}
              value={form.defense}
              onBlur={handleBlur}
            />
            {
              errors.defense && <p style={{ color: 'green' }} >{errors.defense}</p>
            }
            <br></br>
            <br></br>
            <label htmlFor='speed' >Velocidad {`(entre 1-100)`} </label>
            <input
              type='number'
              name='speed'
              min="1"
              max="100"
              id='speed'

              onChange={handleChange}
              value={form.speed}
              onBlur={handleBlur}
            />
            {

              errors.speed && <p style={{ color: 'green' }} >{errors.speed}</p>

            }
            <br></br>
            <br></br>
            <div >
              <label>Tipos de Pokemon </label>
              <select value={form} name='type' onChange={handleChange} >
                <option  >---</option>
                <option value='normal' >Normal</option>
                <option value='fighting' >Fighting </option>
                <option value='flying' >Flying</option>
                <option value='poison' >Poison </option>
                <option value='ground' >Ground</option>
                <option value='rock' >Rock</option>
                <option value='bug'  >Bug </option>
                <option value='ghost'  >Ghost  </option>
                <option value='steel'  >Steel </option>
                <option value='fire' >Fire </option>
                <option value='water'  >Water</option>
                <option value='grass'  >Grass </option>
                <option value='electric'  >Electric</option>
                <option value='psychic'  >Psychic </option>
                <option value='ice'  >Ice </option>
                <option value='dragon'  >Dragon </option>
                <option value='dark'  >Dark  </option>
                <option value='fairy'  >Fairy </option>
                <option value='uknown'  >Unknown </option>
                <option value='shadow'  >Shadow  </option>

              </select>
              {
               errors.type && <p style={{ color: 'green' }} >{errors.type}</p>
              }
              <ul>
                {
                  form.type && form.type.map((el, indice) => (
                    <div key={indice} >
                      <li key={indice} >{el}</li>
                      <button className='btn' key={el} onClick={(e) => { handleEliminate(el) }}>X</button>
                    </div>
                  )
                  )

                }
              </ul>
            </div>
            {
              errors.type && <p style={{ color: 'green' }} >{errors.type[0]}</p>
            }
            <br />
            <label htmlFor='img' >Imagen </label>
            <input
              type='text'
              name='img'
              id='img'
              onChange={handleChange}
              value={form.img}
              onBlur={handleBlur}
            />
            <br />
            {
              errors.img && <p style={{ color: 'green' }} >{errors.img}</p>
            }
            <br></br>
           
            <br></br>
          </div>
          <button className='btn' type='submit'>Crear Pokemon</button>
        </form>
      </div>
    </div>
  )
}
