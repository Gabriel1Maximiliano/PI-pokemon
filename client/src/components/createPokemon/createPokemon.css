.contenedor-formulario{
    background-color: aqua;
    background-image: url('../../images/gobattleleague-announcement-2.jpg');
    background-repeat: no-repeat;
    background-size: cover;


}
.formulario{
    background-color:#F0E7E9;
    width: 50%;
    margin-left: auto;
    margin-right: auto;
    border-radius: 15px;
    padding-top: 15px;
}
.btn{
    border-radius: 6px;
    transition: .5s;
    height: 100%;
    margin-left: 2px;
    font-family: 'Lucida Sans', 'Lucida Sans Regular', 'Lucida Grande', 'Lucida Sans Unicode', Geneva, Verdana, sans-serif; font-weight: 1000;
    font-weight: 1000;
    color: aliceblue;
    background-color: coral;
    text-decoration: none;
  }
  .btn:hover{
    background-color: coral;
    transition: .5s;
    color:black
  }
  .input{
    background-color: #CECEC9;
    transition: .5s;
    border-radius: 5px;
    width: 70%;
    text-align: center;
    
  }
  .decoration-input{
    background-color: red;
    border-radius: 5px;
    width: 46%;
    margin-left: auto;
    margin-right: auto;
    align-items: flex-start;
  }
 