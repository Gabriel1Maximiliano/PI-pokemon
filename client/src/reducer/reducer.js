import {
 
  FILTER_POKE_DB,
  ATTACK_ORDER,
  GET_ALL_POKEMON,
  GET_DETAILS_POKEMON,
  ALPHABETIC_ORDER,
  GET_BY_NAME,
  CREATE_POKEMON,
  BRING_ALL} from '../actionTypes/index';

const initialState = {
    allPokemons:[],
    pokemonDetails:[],
    auxAllPokemons:[],
    aux2AllPokemons:[]
}

const reducer = (state = initialState, { type, payload }) => {
  switch (type) {

  case GET_ALL_POKEMON:
    return{
      ...state,
      allPokemons:payload,
      auxAllPokemons:payload,
      aux2AllPokemons:payload
    }
    case GET_BY_NAME:
      
      return{
         ...state,
         allPokemons: payload
      }
      case GET_DETAILS_POKEMON:
       
      return{
        ...state,
        pokemonDetails:payload[0]
      }
      case CREATE_POKEMON:
        return{
          ...state
        }
        case ALPHABETIC_ORDER:
          
      let alphabeticOrder;
     
      if(payload ==='z-a'){
       
        alphabeticOrder= state.auxAllPokemons.sort(function (a, b) {
          if (a.name > b.name) {
            return -1;
          }
          if (b.name > a.name) {
            return 1;
          }
          return 0;
        })
      }
      else{
        alphabeticOrder= state.auxAllPokemons.sort(function (a, b) {
          if (a.name > b.name) {
            return 1;
          }
          if (b.name > a.name) {
            return -1;
          }
          return 0;
        })
      }
  
          return{
            ...state,
            allPokemons:alphabeticOrder
          }

          case  ATTACK_ORDER:

            
            let strengthSortArray;
           
            if(payload ==='des'){
              
              strengthSortArray=state.allPokemons.sort(function (a, b) {
                if (a.strength > b.strength) {
                  return -1;
                }
                if (b.strenght > a.strength) {
                  return 1;
                }
                return 0;
              })
            }
           
            else{
              strengthSortArray=state.allPokemons.sort(function (a, b) {
              
                if (a.strength > b.strength) {
                  return 1;
                }
                if (b.strength > a.strength) {
                  return -1;       
                }
                return 0;
              })
              
            }
    
            return{
              ...state,
              allPokemons:strengthSortArray
            }
            case FILTER_POKE_DB:
              let filteredPoke=state.auxAllPokemons;
              let filtroAuxiliar = state.aux2AllPokemons;
              
             console.log('entre con un valor'+payload)//
             
               let resp = payload ==='api'?  filteredPoke.filter(p=> !p.createdInDb ):payload ==='all'? state.allPokemons :filtroAuxiliar.filter(p=>p.createdInDb);
               
            return { 
               ...state,
               allPokemons: resp
            }
            case  BRING_ALL:
              return  {
                ...state,
                allPokemons: payload
              }
  default:
    return state
  }
}
export default  reducer;