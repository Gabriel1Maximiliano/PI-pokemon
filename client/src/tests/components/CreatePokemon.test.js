import { render,screen,fireEvent} from "@testing-library/react"
import { Provider } from "react-redux"
import { CreatePokemon } from "../../components/createPokemon/CreatePokemon"
import { MemoryRouter } from "react-router-dom"
import{ createStore } from 'redux';
import reducer from "../../reducer/reducer"
 

const store  = (createStore(reducer))

 

 describe('Pruebas sobre <CreatePokemon />', () => {
    
       test('Deba hacer un match con el snapshot', () => { 

          const { container } = render(
          <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)

            expect( container ).toMatchSnapshot()
        })

        test('Debe contener el texto "Nombre del Pokemon"', () => { 
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)
         expect(screen.getByText('Nombre del Pokemon')).toBeTruthy() 
         })

         test('No debe haber mensajes de error ni  bien se monta el componente', () => {
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)
             expect(screen.queryByText(/'El campo nombre es requerido'/)).not.toBeInTheDocument();
            expect(screen.queryByText(/'Agregue altura a su Pokemon'/)).not.toBeInTheDocument();
            expect(screen.queryByText(/'Agregue peso a su Pokemon'/)).not.toBeInTheDocument();
            expect(screen.queryByText(/'Agregue puntos de vida a su Pokemon'/)).not.toBeInTheDocument();
            expect(screen.queryByText(/  'Agregue fuerza de ataque a tu Pokemon'/)).not.toBeInTheDocument();


         })

         test('Debe mostrar la leyenda "Complete los campos" si hace submit con el formulario vacÃ­o', () => {
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)

           const btnSubmint = screen.getByRole('button', {name: /Crear Pokemon/i});
           window.alert = jest.fn();
             fireEvent.click(btnSubmint)
            let error = window.alert.mock.calls[0];
            expect(error[0]).toEqual('Complete los campos')
         })
       
    });

    describe('Pruebas de mensajes de error en <CreatePokemon />', () => { 
       
        test('Muestra mensaje de error si el campo nombre no esta completado',() => {
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)

            
             fireEvent.blur(screen.getByLabelText(/Nombre del Pokemon/,{value:''}))
                    
           
            expect(screen.getByText(/El campo nombre es requerido/i)).toBeInTheDocument();
         
        })
        test('Muestra mensaje de error en el campo "Altura entre"',() => {
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)

            
             fireEvent.blur(screen.getByLabelText(/Altura entre/,{value:''}))
                    
           
            expect(screen.getByText(/Agregue altura a su Pokemon/i)).toBeInTheDocument();
         
        });
        test('Muestra mensaje de error en el campo " Peso entre"',() => {
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter> 
            </Provider>)

            
             fireEvent.blur(screen.getByLabelText(/Peso entre /,{value:''}))
                    
           
            expect(screen.getByText(/Agregue peso a su Pokemon/i)).toBeInTheDocument();
         
        });
        test('Muestra mensaje de error en el campo "Vida entre"',() => {
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)

            
             fireEvent.blur(screen.getByLabelText(/Vida entre/,{value:''}))
                    
           
            expect(screen.getByText(/Agregue puntos de vida a su Pokemon/i)).toBeInTheDocument();
         
        });
        test('Muestra mensaje de error en el campo "Ataque entre"',() => {
            render(
            <Provider store={store}>
                <MemoryRouter>
                <CreatePokemon />
                </MemoryRouter>
            </Provider>)

            
             fireEvent.blur(screen.getByLabelText(/Ataque entre/,{value:''}))
                    
           
            expect(screen.getByText(/Agregue fuerza de ataque a tu Pokemon/i)).toBeInTheDocument();
         
        });
    
    
     })
    
    
     
    
     //Defensa entre
     //Velocidad
        
        
  
