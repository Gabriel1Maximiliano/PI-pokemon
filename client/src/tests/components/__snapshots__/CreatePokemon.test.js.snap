// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <CreatePokemon /> Deba hacer un match con el snapshot 1`] = `
<div>
  <div
    class="contenedor-formulario"
  >
    <br />
    <div
      class="navigation-bar"
    >
      <a
        class="nav-link"
        href="/home"
      >
        Home
      </a>
      <a
        class="nav-link"
        href="/home/createPokemon"
      >
        Crear Pokemon
      </a>
      <a
        aria-current="page"
        class="nav-link active"
        href="/"
      >
        Landing page
      </a>
    </div>
    <br />
    <div
      class="formulario"
    >
      <form>
        <label
          for="name"
        >
          Nombre del Pokemon
        </label>
        <br />
        <br />
        <div
          class="decoration-input"
        >
          <input
            class="input"
            id="name"
            name="name"
            placeholder="Ingrese nombre del Pokemon"
            type="text"
            value=""
          />
          <br />
          <br />
          <label
            for="height"
          >
            Altura entre 
            (1-30)
             
          </label>
          <input
            id="height"
            max="30"
            min="1"
            name="height"
            type="number"
            value=""
          />
          <br />
          <br />
          <label
            for="weight"
          >
            Peso entre 
            (1-300)
             
          </label>
          <input
            id="weight"
            max="300"
            min="1"
            name="weight"
            type="number"
            value=""
          />
          <br />
          <br />
          <label
            for="hp"
          >
            Vida entre 
            (1-100)
             
          </label>
          <input
            id="hp"
            max="100"
            min="1"
            name="hp"
            type="number"
            value=""
          />
          <br />
          <br />
          <label
            for="strength"
          >
            Ataque entre 
            (1-300)
             
          </label>
          <input
            id="strength"
            max="300"
            min="1"
            name="strength"
            type="number"
            value=""
          />
          <br />
          <br />
          <label
            for="defense"
          >
            Defensa entre 
            (1-100)
             
          </label>
          <input
            id="defense"
            max="100"
            min="1"
            name="defense"
            type="number"
            value=""
          />
          <br />
          <br />
          <label
            for="speed"
          >
            Velocidad 
            (entre 1-100)
             
          </label>
          <input
            id="speed"
            max="100"
            min="1"
            name="speed"
            type="number"
            value=""
          />
          <br />
          <br />
          <div>
            <label>
              Tipos de Pokemon 
            </label>
            <select
              name="type"
            >
              <option>
                ---
              </option>
              <option
                value="normal"
              >
                Normal
              </option>
              <option
                value="fighting"
              >
                Fighting 
              </option>
              <option
                value="flying"
              >
                Flying
              </option>
              <option
                value="poison"
              >
                Poison 
              </option>
              <option
                value="ground"
              >
                Ground
              </option>
              <option
                value="rock"
              >
                Rock
              </option>
              <option
                value="bug"
              >
                Bug 
              </option>
              <option
                value="ghost"
              >
                Ghost  
              </option>
              <option
                value="steel"
              >
                Steel 
              </option>
              <option
                value="fire"
              >
                Fire 
              </option>
              <option
                value="water"
              >
                Water
              </option>
              <option
                value="grass"
              >
                Grass 
              </option>
              <option
                value="electric"
              >
                Electric
              </option>
              <option
                value="psychic"
              >
                Psychic 
              </option>
              <option
                value="ice"
              >
                Ice 
              </option>
              <option
                value="dragon"
              >
                Dragon 
              </option>
              <option
                value="dark"
              >
                Dark  
              </option>
              <option
                value="fairy"
              >
                Fairy 
              </option>
              <option
                value="uknown"
              >
                Unknown 
              </option>
              <option
                value="shadow"
              >
                Shadow  
              </option>
            </select>
            <ul />
          </div>
          <br />
          <label
            for="img"
          >
            Imagen 
          </label>
          <input
            id="img"
            name="img"
            type="text"
            value=""
          />
          <br />
          <br />
          <br />
        </div>
        <button
          class="btn"
          type="submit"
        >
          Crear Pokemon
        </button>
      </form>
    </div>
  </div>
</div>
`;
